engine/src/main/java/org/hibernate/validator/internal/engine/path/PublicPath.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright Red Hat Inc. and Hibernate Authors
+ */
+package org.hibernate.validator.internal.engine.path;
+
+import java.io.Serial;
+import java.io.Serializable;
+import java.util.Collections;
+import java.util.Iterator;
+
+import jakarta.validation.Path;
+
+final class PublicPath implements Path, Serializable {
+
+	@Serial
+	private static final long serialVersionUID = -8906501301223202169L;
+
+	private final NodeImpl leafNode;
+	private final NodeImpl[] nodes;
+	private final int hashCode;
+
+	PublicPath(ModifiablePath path) {
+		this.leafNode = path.getLeafNode();
+		this.nodes = NodeImpl.constructPath( this.leafNode );
+		this.hashCode = path.hashCode();
```

Comment:
I think I would avoid doing that here, and let it be done later.

Except if we are using it on the HV side or if it's a common use case amongst consumers?